### DISCUSSION TITLE
Application-Defined SQLite Functions

### DISCUSSION THREAD

**Discussion by Toromyx** [SeaQL/sea-orm#1666] [Ideas]
I would like to suggest implementing a developer-friendly interface to create application-defined SQLite functions.

SQLite provides the functionality to create application-defined functions which just run code in the same binary since SQLite is packaged in it. The documentation can be found under <https://www.sqlite.org/appfunc.html>.

The `libsqlite3_sys` functions can be found here: <https://docs.rs/libsqlite3-sys/latest/libsqlite3_sys/?search=create_function>

I tried implementing this myself but quickly realized that it is a pretty large task. Then I looked for other projects which might have implemented this and stumbeld upon the implementation in `diesel`: <https://github.com/diesel-rs/diesel/pull/1691>. This is an old PR, the new code is slighty different. The developer-facing function can be found in `diesel::sqlite::connection::SqliteConnection::register_sql_function`: <https://github.com/diesel-rs/diesel/blob/master/diesel/src/sqlite/connection/mod.rs#L317> (The function is hidden in the documentation).

Then I realized that instead of implementing it myself, it would be better to have my ORM implement it. I looked for issues and discussions already mentioning custom functions and only found <https://github.com/SeaQL/sea-orm/discussions/687>. My suggested feature would be one solution to the problem mentioned there.

Do you think this feature request fits this project?

**Comment by billy1624** [comment]
Hey @Toromyx, I afraid you need to write raw SQL in your migration file https://www.sea-ql.org/SeaORM/docs/migration/writing-migration/#raw-sql

A PR would be welcomed! Let me know if you need my help :)

**Reply by Toromyx** [reply]
Thank you for the answer, but I'm not talking about the `CREATE FUNCTION` statement in MySQL or something written in SQL. I'm talking about functions written in Rust and called from SQLite. This is only possible in SQLite since that database engine is compiled together with the rest of the application code in one binary.

**Reply by billy1624** [reply]
I'm not familiar with that approach. But I think it's doable in SeaORM. We'd need to introduce a new backend (https://github.com/rusqlite/rusqlite) and implements SeaORM traits accordingly.

**Reply by Toromyx** [reply]
Maybe instead of introducing a new backend, an issue on https://github.com/launchbadge/sqlx could be raised to implement it on their side? There probably was a reason sqlx was chosen over rusqlite?

There already seems to be an abandoned discussion: https://github.com/launchbadge/sqlx/discussions/1418

**Reply by billy1624** [reply]
We chose SQLx because it support multiple database backends

**Reply by Toromyx** [reply]
Okay. Once I find the time, I will try to implement a rusqlite backend for sea-orm.

**Reply by Toromyx** [reply]
I'm back at the point where I want to implement this. I found this discussion which implies that using rusqlite ist not a good idea: <https://github.com/SeaQL/sea-orm/discussions/721>.
I will think about how I want to implement this and create an issue or PR here or against sqlx.

**Comment by Toromyx** [comment]
For anyone coming here: I did some experiments with <https://github.com/asg017/sqlite-loadable-rs> and <https://github.com/CGamesPlay/sqlite3_ext> with the goal of using <https://www.sqlite.org/c3ref/auto_extension.html> to load an SQLite extension directly via <https://github.com/rusqlite/rusqlite/tree/master/libsqlite3-sys>, completely sidestepping any ORM.
But I found these libraries to not be ready yet for static linking with a current version of libsqlite3-sys. I might revisit them.
